internal/failure_converter.go

```diff
@@ -0,0 +1,269 @@
+		}
+		failure.FailureInfo = &failurepb.Failure_ApplicationFailureInfo{ApplicationFailureInfo: failureInfo}
+	case *CanceledError:
+		failureInfo := &failurepb.CanceledFailureInfo{
+			Details: convertErrDetailsToPayloads(err.details, dc),
+		}
+		failure.FailureInfo = &failurepb.Failure_CanceledFailureInfo{CanceledFailureInfo: failureInfo}
+	case *PanicError:
+		failureInfo := &failurepb.ApplicationFailureInfo{
+			Type: getErrType(err),
+		}
+		failure.FailureInfo = &failurepb.Failure_ApplicationFailureInfo{ApplicationFailureInfo: failureInfo}
+		failure.StackTrace = err.StackTrace()
+	case *workflowPanicError:
+		failureInfo := &failurepb.ApplicationFailureInfo{
+			Type:         getErrType(&PanicError{}),
+			NonRetryable: true,
+		}
+		failure.FailureInfo = &failurepb.Failure_ApplicationFailureInfo{ApplicationFailureInfo: failureInfo}
+		failure.StackTrace = err.StackTrace()
+	case *TimeoutError:
+		failureInfo := &failurepb.TimeoutFailureInfo{
+			TimeoutType:          err.timeoutType,
+			LastHeartbeatDetails: convertErrDetailsToPayloads(err.lastHeartbeatDetails, dc),
+		}
+		failure.FailureInfo = &failurepb.Failure_TimeoutFailureInfo{TimeoutFailureInfo: failureInfo}
+	case *TerminatedError:
+		failureInfo := &failurepb.TerminatedFailureInfo{}
+		failure.FailureInfo = &failurepb.Failure_TerminatedFailureInfo{TerminatedFailureInfo: failureInfo}
+	case *ServerError:
+		failureInfo := &failurepb.ServerFailureInfo{
+			NonRetryable: err.nonRetryable,
+		}
+		failure.FailureInfo = &failurepb.Failure_ServerFailureInfo{ServerFailureInfo: failureInfo}
+	case *ActivityError:
+		failureInfo := &failurepb.ActivityFailureInfo{
+			ScheduledEventId: err.scheduledEventID,
+			StartedEventId:   err.startedEventID,
+			Identity:         err.identity,
+			ActivityType:     err.activityType,
+			ActivityId:       err.activityID,
+			RetryState:       err.retryState,
+		}
+		failure.FailureInfo = &failurepb.Failure_ActivityFailureInfo{ActivityFailureInfo: failureInfo}
+	case *ChildWorkflowExecutionError:
+		failureInfo := &failurepb.ChildWorkflowExecutionFailureInfo{
+			Namespace: err.namespace,
+			WorkflowExecution: &commonpb.WorkflowExecution{
+				WorkflowId: err.workflowID,
+				RunId:      err.runID,
+			},
+			WorkflowType:     &commonpb.WorkflowType{Name: err.workflowType},
+			InitiatedEventId: err.initiatedEventID,
+			StartedEventId:   err.startedEventID,
+			RetryState:       err.retryState,
+		}
+		failure.FailureInfo = &failurepb.Failure_ChildWorkflowExecutionFailureInfo{ChildWorkflowExecutionFailureInfo: failureInfo}
+	default: // All unknown errors are considered to be retryable ApplicationFailureInfo.
+		failureInfo := &failurepb.ApplicationFailureInfo{
+			Type:         getErrType(err),
+			NonRetryable: false,
+		}
+		failure.FailureInfo = &failurepb.Failure_ApplicationFailureInfo{ApplicationFailureInfo: failureInfo}
+	}
+
+	failure.Cause = InternalErrorToFailure(errors.Unwrap(err), dc)
```

Comment:
This does not properly encode nested attributes.

Instead of walking the causes after this is done and to avoid these mistakes in the codebase in general, lets get rid of these two top-level functions and put them _inside_ the methods of the default converter itself. If someone wants to call these from their converter, they can call `GetDefaultFailureConverter().ErrorToFailure` or `GetDefaultFailureConverter().FailureToError` to get default behavior.

Reply:
ah good catch, i'll refactor